---
title: 'Colon Adenocarcinoma (COAD) Exploratory Data Analysis'
author: "Sehyun Oh, Britney Pheng"
date: "`r format(Sys.time(), '%B %d, %Y')`"
format:
    html:
        fontsize: 14pxs
        toc: true
        top-depth: 3
output: html_document
---

# Initial Setup
## Load packages
```{r packages, warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  # BiocManager
  library(GenomicSuperSignature)
  library(curatedTCGAData)
  library(MultiAssayExperiment)
  library(TCGAutils)
  library(ComplexHeatmap)
  
  # CRAN
  library(tidyverse) # includes dplyr, ggplot2, magrittr, tidyr
  library(magick)
  library(wordcloud)
  library(ztable)
  library(metafolio)
})
```

## Load RAVmodel
```{r message=FALSE, warning=FALSE}
RAVmodel <- getModel('C2', load=TRUE)
```
## TCGA-COAD RNAseq data
### From GSEABenchmarkeR
```{r eval=FALSE}
# Raw read counts from GSE62944 from ExperimentHub
tcga <- GSEABenchmarkeR::loadEData("tcga",
                                   cache = FALSE,
                                   paired = FALSE,
                                   map2entrez = FALSE)
assay(tcga$COAD) <- log2(assay(tcga$COAD) + 1)
tcga_coad <- tcga$COAD # 524 samples
save(tcga_coad, file = "inst/extdata/TCGA_COAD_validationDatase.rda")
```

## Select COAD RNA metadata
```{r message=FALSE, warning=FALSE}
coad <- curatedTCGAData(diseaseCode = 'COAD',
                        assays = 'RNA*',
                        version = '2.0.1',
                        dry.run = FALSE)

coad_rna <- getWithColData(coad,
                           'COAD_RNASeq2Gene-20160128',
                           mode = 'append')
coad_meta <- colData(coad_rna)
```

# heatmapTable: COAD
```{r message=FALSE}
validate_coad <- validate(tcga_coad, RAVmodel)
heatmapTable(validate_coad, RAVmodel)
```

```{r message=FALSE}
assay(coad_rna) <- log2(assay(coad_rna) + 1)

validate_coad_rna <- validate(coad_rna, RAVmodel)
heatmapTable(validate_coad_rna, RAVmodel)
```

# Subset
## Filter attributes
```{r}
sparsity_summary <- table(colSums(is.na(coad_meta)))
sparsity_summary
```

## Sparsity Plot
```{r echo=FALSE}
plot(stack(sparsity_summary)$ind,
     stack(sparsity_summary)$values)
```

```{r}
# Select columns with >10% completeness
keep_attribute_ind <- which(colSums(!is.na(coad_meta)) > round(nrow(coad_meta)/10))
meta_sub1 <- coad_meta[keep_attribute_ind]
meta_sub1 <- subset(meta_sub1, select= -patientID)
```

```{r}
# Randomly select for 100 rows
set.seed(1)
random_sample_ind <- sample(1:nrow(meta_sub1), 100)
meta_sub2 <- meta_sub1[random_sample_ind,]
```

```{r}
# Check for data types in listData
unique(sapply(coad_meta@listData, type))

charcTb <- meta_sub2[, sapply(meta_sub1, class) == 'character']
numTb <- meta_sub2[, sapply(meta_sub1, class) %in% c('numeric', 'integer')]
```

```{r}
# Calculate validation scores
sampleScore <- calculateScore(coad_rna, RAVmodel)
```

```{r}
validated_ind <- validatedSignatures(validate_coad_rna, RAVmodel, num.out = 15, scoreCutoff = 0.45, indexOnly = TRUE)

# Subset sampleScore to join with MCPcounter
sampleScore_sub <- sampleScore[random_sample_ind, validated_ind] %>% as.data.frame()
```

# Calculate R-Squared Value for Numeric Variables
```{r calculateRsq}
# R squared value function
calculateRsq <- function (x, y) stats::cor(x, y, use = 'na.or.complete') ^ 2
```

```{r warning=FALSE}
# Calculate r-squared for numeric attributes
rsq_numAttr <- as.data.frame(matrix(nrow = ncol(numTb),
                                    ncol = ncol(sampleScore_sub)))

colnames(rsq_numAttr) <- colnames(sampleScore_sub)
rownames(rsq_numAttr) <- colnames(numTb)

for (i in seq_len(ncol(numTb))) {
  for (j in seq_len(ncol(sampleScore_sub))) {
    rsq <- calculateRsq(numTb[, i], sampleScore_sub[, j])
    rsq_numAttr[i, j] <- rsq
  }
}

rsq_numAttr <- na.omit(rsq_numAttr)
```

```{r}
# Remove batch effect variables
batch_num_ind <- grep('analyte|portion|procurement|aliquot|uuid|barcode',
                  rownames(rsq_numAttr))
rsq_numAttr_no_batch <- rsq_numAttr[-batch_num_ind,]
```

```{r}
max_rav <- apply(rsq_numAttr_no_batch, 1, max)
max_attr <- which(max_rav > 0.5)

target_rsq <- rsq_numAttr_no_batch[max_attr,]
```

# heatmapTable
```{r results='asis'}
options(ztable.type='html')
z = ztable(target_rsq)
z %>% makeHeatmap(palette="Blues")
```

```{r}
heatmap(as.matrix(target_rsq), scale="column")
```
# Separate out Factor Variables by number of levels (1, 2, 3+) not including NA
```{r}
# Convert to factor data type
factorTb <- meta_sub2[, sapply(meta_sub1, class) == 'character']

factorTb[sapply(factorTb, is.character)] <- lapply(factorTb[sapply(factorTb, is.character)], factor, exclude = NULL)

#any(is.na(levels(factorTb[,2])))
#!any(is.na(levels(factorTb[,6])))

single_fac_ind <- c()
binary_fac_ind <- c()
multi_fac_ind <- c()

# Testing factor grouping
for (i in 1:length(factorTb)) {
  if (nlevels(factorTb[,i]) == 1 | 
      (nlevels(factorTb[,i]) == 2 & 
       any(is.na(levels(factorTb[,i]))))
      ) {
    single_fac_ind <- c(single_fac_ind, i)
  } else if (#nlevels(factorTb[,i]) == 3 & 
             #any(is.na(levels(factorTb[,i]))) |
             
             (nlevels(factorTb[,i]) == 2 & 
              !any(is.na(levels(factorTb[,i]))))
          ) {
    binary_fac_ind <- c(binary_fac_ind, i)
  } else {
    multi_fac_ind <- c(multi_fac_ind, i)
  }
}

new_factorTb <- factorTb[,multi_fac_ind]
binary_factor <- factorTb[,binary_fac_ind]
binary_factor <- binary_factor[-c(7,8)]
single_factor <- factorTb[,single_fac_ind]
```

# Calculate F-statistic (ANOVA) for Character Variables
```{r}
aov_res <- as.data.frame(matrix(nrow = ncol(new_factorTb),
                                ncol = ncol(sampleScore_sub)))

rownames(aov_res) <- colnames(new_factorTb)
colnames(aov_res) <- colnames(sampleScore_sub)

aov_coad_fvalue <- aov_res
aov_coad_pvalue <- aov_res

for (i in seq_len(ncol(sampleScore_sub))) {
  for (j in seq_len(ncol(new_factorTb))) {
    
    ## ANOVA
    aov <- aov(sampleScore_sub[, i] ~ new_factorTb[, j])
    
    ## F-statistic
    fval <- summary(aov)[[1]]$`F value`[1]
    aov_coad_fvalue[j, i] <- fval
    
    ## p-value
    pval <- summary(aov)[[1]]$`Pr(>F)`[1]
    aov_coad_pvalue[j, i] <- pval
  }
}
```

```{r}
# Select for p-values < 0.01
min_rav <- apply(aov_coad_pvalue, 1, min)
min_attr <- which(min_rav < 0.01)

target_coad_aov_fvalue <- aov_coad_fvalue[min_attr,]
target_coad_aov_pvalue <- aov_coad_pvalue[min_attr,]
```

```{r}
batch_char_ind <- grep('analyte|portion|procurement|aliquot|uuid|barcode',
                       rownames(target_coad_aov_fvalue))
coad_aov_fvalue <- target_coad_aov_fvalue[-batch_char_ind,]
coad_aov_pvalue <- target_coad_aov_pvalue[-batch_char_ind,]
```

```{r}
heatmap(as.matrix(coad_aov_fvalue), main = 'COAD F-Statistics')
```

```{r}
sig_fval <- as.data.frame(matrix(nrow = ncol(new_factorTb),
                                ncol = ncol(sampleScore_sub)))

rownames(sig_fval) <- colnames(new_factorTb)
colnames(sig_fval) <- colnames(sampleScore_sub)

for (i in seq_len(ncol(sampleScore_sub))) {
  for (j in seq_len(ncol(new_factorTb))) {
    
    if (!is.null(summary(aov(sampleScore_sub[, i] ~ new_factorTb[, j]))[[1]]$`Pr(>F)`[1])) {
      
      if (!is.null(summary(aov(sampleScore_sub[, i] ~ new_factorTb[, j]))[[1]]$`F value`[1]) &
          (summary(aov(sampleScore_sub[, i] ~ new_factorTb[, j]))[[1]]$`Pr(>F)`[1] < 0.05)) {
        sig_fval[j, i] <- summary(aov(sampleScore_sub[, i] ~ new_factorTb[, j]))[[1]]$`F value`[1]
      } else {
        next
      }
    } else {
      next
    }
  }
}

fstat_ind <- c()

for (i in seq_len(nrow(sig_fval))) {
  if (sum(is.na(sig_fval[i, ])) < 220) {
    fstat_ind <- c(fstat_ind, i)
  } else {
    next
  }
}

sig_fval <- sig_fval[fstat_ind,]
batch_char_ind_2 <- grep('analyte|portion|procurement|aliquot|uuid|barcode',
                  rownames(sig_fval))
sig_fval <- sig_fval[-batch_char_ind_2,]

```

```{r results='asis', message=FALSE}
options(ztable.type='html')
#ztable(sig_fval) %>%
 # makeHeatmap(palette = 'Blues') #%>%
  #print(caption='ANOVA F-stats with p-values < 0.01')
```

# Calculate t-test for binomial Character Variables
```{r}
#Remove factors without enough y-observations
for (i in seq_len(ncol(binary_factor))) {
  x <- summary(binary_factor[,i])
  
  if (all(x > 1)) {
    print(x)
  } else {
    print(paste("index", i, "does not have enough observations"))
  }
}

bin_fac <- binary_factor[,-c(5, 12, 26)]

t_test_res <- as.data.frame(matrix(nrow = ncol(bin_fac),
                                ncol = ncol(sampleScore_sub)))

rownames(t_test_res) <- colnames(bin_fac)
colnames(t_test_res) <- colnames(sampleScore_sub)

ttest_coad_tvalue <- t_test_res
ttest_coad_pvalue <- t_test_res

for (i in seq_len(ncol(sampleScore_sub))) {
  for (j in seq_len(ncol(bin_fac))) {
    
    ## t-test
      t_test <- t.test(sampleScore_sub[, i] ~ bin_fac[,j])
    
    ## t-value
      tval <- t_test$statistic
      ttest_coad_tvalue[j, i] <- tval
    
    ## p-value
      pval <- t_test$p.val
      ttest_coad_pvalue[j, i] <- pval
  }
}

batch_char_ind <- grep('analyte|portion|procurement|aliquot|uuid|barcode',
                  rownames(ttest_coad_tvalue))
ttest_coad_tvalue_2 <- ttest_coad_tvalue[-batch_char_ind,]
```

```{r}
heatmap(as.matrix(ttest_coad_tvalue_2), main = 'COAD T-Test')
```

```{r}
#Only view t-values with significant p-value (<0.05)
ttest_coad_sig_tvalue <- t_test_res


 for (i in seq_len(ncol(sampleScore_sub))) {
   for (j in seq_len(ncol(bin_fac))) {
     if (t.test(sampleScore_sub[, i] ~ bin_fac[,j])$p.val < 0.05) {
       ttest_coad_sig_tvalue[j, i] <- t.test(sampleScore_sub[, i] ~ bin_fac[,j])$statistic
     }
   }
 }

batch_char_ind <- grep('analyte|portion|procurement|aliquot|uuid|barcode',
                  rownames(ttest_coad_sig_tvalue))
ttest_coad_sig_tvalue <- ttest_coad_sig_tvalue[-batch_char_ind,]
```

```{r results='asis', message=FALSE}
options(ztable.type='html')
ztable(ttest_coad_sig_tvalue) %>%
  makeHeatmap(palette = 'Blues') %>%
  print(caption='T-test with p-values < 0.05')
```

## RAV Exploration
```{r}
ind <- 833
findStudiesInCluster(RAVmodel, ind, studyTitle = TRUE)
drawWordcloud(RAVmodel, ind)
```

```{r}
## Function to Display Categorical Attributes' Score Plots
 
graph_categorical <- function(a, b, c, d = 'Legend') {
  sampleScore1 <- paste0('RAV', a)
  sampleScore2 <- paste0('RAV', b)

  plot_data <- sampleScore_sub[, c(sampleScore1, sampleScore2)]
  colnames(plot_data)[1] <- "sampleScore1"
  colnames(plot_data)[2] <- "sampleScore2"

  plot_data <- data.frame(plot_data, new_factorTb)
  plot_data_2 <- plot_data[which(!is.na(factor(plot_data[[c]]))),]

  colors <- gg_color_hue(length(unique(plot_data_2[[c]])))
  colors.toplot <- c(colors)

  pA <- ggplot(plot_data_2,
             aes(x = sampleScore1, y = sampleScore2, color = plot_data_2[[c]]
                 )) +
        geom_point() +
        #ggtitle('Score Plot for Gender') +
        scale_color_manual(values = colors.toplot, name = d) +
        theme_bw() +
        theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
        geom_hline(yintercept = 0, linetype = 'dashed') +
        geom_vline(xintercept = 0, linetype = 'dashed') +
        xlab(sampleScore1) + ylab(sampleScore2)

print(pA)
}
```

```{r warning=FALSE}
graph_categorical(834, 833, 'pathology_M_stage', 'Pathology M Stage')
```
```{r warning=FALSE}
graph_categorical(834, 1302, 'patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status', 'Microsatellite')
```

```{r warning=FALSE}
graph_categorical(834, 517, 'patient.microsatellite_instability', new_factorTb, 'Microsatellite Instability')
```


```{r warning=FALSE}
graph_categorical(832, 833, 'race', new_factorTb, 'Race')
```


